packages/frontend/component/src/ui/modal/modal.tsx

```diff
@@ -76,11 +81,44 @@ export const Modal = forwardRef<HTMLDivElement, ModalProps>(
     },
     ref
   ) => {
+    const [{ status }, toggle] = useTransition({
+      timeout: animationTimeout,
+    });
+    useEffect(() => {
+      toggle(customOpen);
+    }, [customOpen]);
+
+    const handleOpenChange = useCallback(
+      (open: boolean) => {
+        if (!open) {
+          toggle(open);
+          const timeout = setTimeout(() => {
```

Comment:
It's a little strange to handle timeout ourself, will it be better to handle by `react-transition-state`? Maybe something like `onStateChange`. What do you think about @pengx17 ?

Reply:
I think the reason why adding timeout here is to make sure the content wrapped in the modal will still show during exiting when closing is triggered by modal itself like clicking outside. However the issue still happens when the upstream changes changes `open` to false.

The fix here is not complete. Maybe you can just skip this patch - or using web transition api instead.

Reply:
Not sure if it's okay this way:

```tsx
const [{}, toggle] = useTransition({
  timeout: animationTimeout,
  onStateChange: useCallback(({ current }) => {
    if (current.status === 'exited') customOnOpenChange?.(false)
    if (current.status === 'entered') customOnOpenChange?.(true)
  }, []),
})

<Dialog.Root
  onOpenChange={toggle}
/>
```

Reply:
yeah, better add an animation started/finished callbacks instead of customOnOpenChange

Reply:
Both `onOpenChange` & `onOpenChanged`, but may need to migrate all used.

Reply:
@JimmFly IMO trying to encapsulate content's view transition state (during exiting) inside of this modal dialog seems too tricky without view transition api, since there is no better way to do it when upstream changes open to false.